src/doorman/Channel.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include "Fd.hpp"
+#include "Types.hpp"
+
+#include <cstdint>
+#include <gsl/span>
+#include <netinet/in.h>
+#include <string>
+#include <unistd.h>
+#include <vector>
+
+class Doorman;
+class Xania;
+
+class Channel {
+    int32_t id_{}; // the channel id
+    Doorman *doorman_{}; // one day you will be a reference
+    Xania *mud_{}; // one day you will be a reference
+    Fd fd_;
+    std::string hostname_;
+    std::vector<byte> buffer_;
+    uint16_t port_{};
+    uint32_t netaddr_{};
+    int width_{}, height_{}; /* Width and height of terminal */
+    bool ansi_{}, got_term_{};
+    bool echoing_{};
+    bool sent_reconnect_message_{};
+    bool connected_{}; /* Socket is ready to connect to Xania */
+    pid_t host_lookup_pid_{}; /* PID of look-up process */
+    Fd host_lookup_fd_; // Receive end of a pipe to the look-up process
+    std::string auth_char_name_; /* name of authorized character */
+
+    void async_lookup_process(sockaddr_in address, Fd reply_fd);
+
+public:
+    [[nodiscard]] int32_t id() const { return id_; }
+    void newConnection(Fd fd, sockaddr_in address);
+    void closeConnection();
+    void sendInfoPacket() const;
+    void sendConnectPacket();
+    bool incomingData(gsl::span<const byte> incoming_data);
+    void incomingHostnameInfo();
+    void sendTelopts() const;
+
+    void on_auth(std::string_view name) { auth_char_name_ = name; }
+    [[nodiscard]] const std::string &authed_name() const { return auth_char_name_; }
+    void on_reconnect_attempt();
+    void mark_disconnected() noexcept { connected_ = false; }
+    void send_to_client(gsl::span<const char> span) const { send_to_client(span.data(), span.size_bytes()); }
+    void send_to_client(gsl::span<const byte> span) const { send_to_client(span.data(), span.size_bytes()); }
+    void send_to_client(const void *data, size_t length) const {
+        if (fd_.is_open())
+            fd_.write(data, length);
+    }
+
+    // Intention here is to prevent direct access to the fd.
+    [[nodiscard]] bool has_fd(int fd) const { return fd == fd_.number(); }
+    [[nodiscard]] bool is_closed() const { return !fd_.is_open(); }
+
+    void set_echo(bool echo);
+    void close();
+
+    [[nodiscard]] int set_fds(fd_set &input_fds, fd_set &exception_fds) noexcept {
```

Comment:
I note a mix of implementation in the header vs the cpp file. set_dfs() isn't a one-liner.  Do you use a particular rule for that? e.g. simple pure getters can go in the header, or if it's a template class they all have to.

Reply:
MOSTLY if it's trivial I put it in the header. This is an example of something that grew and I lazily didn't move it. With modern compilers, they can inline things that aren't even in the header, and I tend to err on putting more things in the cpp file I'll move this one as it's out of place.

There's a big movement in the C++ world to put _everything_ in headers. I don't subscribe to this viewpoint. I'm a grumpy chap about it:)